core/arch/arm/plat-hikey/spi_test.c

```diff
@@ -0,0 +1,98 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <drivers/pl022_spi.h>
+#include <drivers/pl061_gpio.h>
+#include <hikey_peripherals.h>
+#include <stdint.h>
+#include <trace.h>
+
+/*
+ * This runs a loopback test by default, so the SPI module will just
+ * receive what is transmitted, i.e. 0x01, 0x80, 0x00.
+ *
+ * In non-loopback mode, the transmit value will ellicit a readback of
+ * the measured value from the ADC chip on the Linksprite 96Boards
+ * Mezzanine card [1], which can be connected to either a sliding
+ * rheostat [2] or photoresistor [3].
+ *
+ * [1] http://linksprite.com/wiki/index.php5?title=Linker_Mezzanine_card_for_96board
+ * [2] http://learn.linksprite.com/96-board/sliding-rheostat
+ * [3] http://learn.linksprite.com/96-board/photoresistor
+ */
+void spi_test(void)
+{
+	struct pl061_data platform_pl061_data;
+	struct pl022_data platform_pl022_data;
+	vaddr_t gpio6_base = peripherals_base(GPIO6_BASE);
+	vaddr_t spi_base = peripherals_base(SPI_BASE);
+	uint8_t tx[3] = {0x01, 0x80, 0x00};
+	uint8_t rx[3] = {0};
+	size_t i, j, num_rxpkts, len = 3;
+
+	DMSG("gpio6_base: 0x%" PRIxVA "\n", gpio6_base);
+	DMSG("spi_base: 0x%" PRIxVA "\n", spi_base);
+
+	DMSG("configure GPIO\n");
+	pl061_init(&platform_pl061_data);
+	pl061_register(gpio6_base, 6);
+
+	DMSG("enable software mode control for chip select\n");
+	pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
+
+	DMSG("mask/disable interrupt for chip select\n");
+	platform_pl061_data.chip.ops->set_interrupt
+		(GPIO6_2, GPIO_INTERRUPT_DISABLE);
+
+	DMSG("configure SPI\n");
+	platform_pl022_data.gpio = &platform_pl061_data.chip;
+	platform_pl022_data.base = spi_base;
+	platform_pl022_data.cs_gpio_base = gpio6_base;
+	platform_pl022_data.clk_hz = SPI_CLK_HZ,
+	platform_pl022_data.speed_hz = SPI_500_KHZ,
+	platform_pl022_data.cs_gpio_pin = GPIO6_2,
+	platform_pl022_data.mode = SPI_MODE0,
+	platform_pl022_data.data_size_bits = 8,
+	platform_pl022_data.loopback = true,
+
+	pl022_configure(&platform_pl022_data);
+	pl022_start(&platform_pl022_data);
+
+	for (j = 0; j < 20; j++) {
+		DMSG("SPI test loop: %zu\n", j);
+		platform_pl022_data.chip.ops->txrx8
+			(&platform_pl022_data.chip, tx, rx, len, &num_rxpkts);
+		for (i = 0; i < num_rxpkts; i++)
+			DMSG("rx[%zu] = 0x%x\n", i, rx[i]);
+
+		/* sleep ~1-2s */
+		for (i = 0; i < 100000000; i++)
+			;
```

Comment:
Using `tee_time_wait()` leads to a translation fault
```
ERROR:   TEE-CORE: 
ERROR:   TEE-CORE: core data-abort at address 0x0 (translation fault)
ERROR:   TEE-CORE:  esr 0x96000046  ttbr0 0x3f079000   ttbr1 0x00000000   cidr 0x0
ERROR:   TEE-CORE:  cpu #0          cpsr 0x200003c4
ERROR:   TEE-CORE: x0  0000000000000000 x1  0000000000000000
ERROR:   TEE-CORE: x2  0000000000000080 x3  0000000000000000
ERROR:   TEE-CORE: x4  000000003f0795fa x5  00000000fffffffe
ERROR:   TEE-CORE: x6  0000000000000002 x7  0000000000000020
ERROR:   TEE-CORE: x8  0000000000000078 x9  000000003f05bd88
ERROR:   TEE-CORE: x10 0000000000000000 x11 00000000f7000000
ERROR:   TEE-CORE: x12 0000000000000000 x13 0000000000000000
ERROR:   TEE-CORE: x14 0000000000000000 x15 0000000000000000
ERROR:   TEE-CORE: x16 0000000000000000 x17 0000000000000000
ERROR:   TEE-CORE: x18 0000000000000000 x19 0000000000000000
ERROR:   TEE-CORE: x20 000000003f0797e0 x21 0000000000000001
ERROR:   TEE-CORE: x22 0000000000000000 x23 0000000000000005
ERROR:   TEE-CORE: x24 000000003f034b37 x25 000000003f079878
ERROR:   TEE-CORE: x26 0000000000000000 x27 0000000000000000
ERROR:   TEE-CORE: x28 0000000000000000 x29 000000003f079770
ERROR:   TEE-CORE: x30 000000003f006e0c elr 000000003f02eb0c
ERROR:   TEE-CORE: sp_el0 000000003f079770
ERROR:   TEE-CORE: Panic
```
I think the test is executed, before you can use `kernel/tee_time` stuff

By the way, my logic analyzer measured 17,8ms for the 100M cycles in this for-loop.